Suppress suggestions while span is in external library

xizheyin opened the pull request:
Fixes rust-lang/rust#139050
Fixes rust-lang/rust#139051 
Fixes rust-lang/rust#139246 
Fixes rust-lang/rust#139247 
Fixes rust-lang/rust#139251 
Fixes rust-lang/rust#139830

xizheyin committed:
Add ui test

Signed-off-by: xizheyin <xizheyin@smail.nju.edu.cn>

fmease commented on compiler/rustc_trait_selection/src/error_reporting/traits/suggestions.rs:
```diff
@@ -449,6 +449,10 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
         err: &mut Diag<'_>,
         trait_pred: ty::PolyTraitPredicate<'tcx>,
     ) -> bool {
+        if obligation.cause.span.from_expansion() {
```
I wonder if `.in_external_macro(self.infcx.tcx.sess.source_map())` would be a better fit?

xizheyin replied:
I think so.Thank you. That makes sense because we can't change only the macros in the external library.

xizheyin committed:
Suppress suggestions on deref in macro when == is in extern macro

Signed-off-by: xizheyin <xizheyin@smail.nju.edu.cn>

xizheyin commented:
> #### Werid Point
>  In the first commit, I use . /x test tests/ui could not reproduce the issue in issue, however I was able to reproduce it using the compiled out version of stage1 in my computer.

I've changed the code. But the strange points are still not solved, does it matter? @fmease 

r? @fmease

fmease commented:
> But the strange points are still not solved

The difference you're likely seeing is the availability vs. unavailability of `std`'s source code. I don't quite remember the reasons as to why the sources of the stage1(?) sysroot aren't available for `tests/ui/` tests but in any case whenever they aren't available, some kinds of subdiagnostics simply won't get shown by the diagnostic emitter IIRC.

---

Regarding the issues you're addressing with this PR as well as with #139399 and #139403, I'm currently leaning towards suppressing structured suggestions wholesale if they point into an external macro, rather than sprinkling these checks all over the compiler.

What do you think, @rust-lang/wg-diagnostics?

The downside of hiding those suggestions: The end user might actually have control over the external crate (multi-crate projects are quite common.

---

I'm also asking myself whether issues #139251, #139253, #139050 etc. are recent(?) regressions? And if so, whether they were intentional. We might want to bisect them (e.g., with cargo-bisect-rustc).

xizheyin commented:
I agree with you, because there are many such cases, it seems that it is a good idea to comprehensively suppress this behavior in diagnostic suggestions.

For the scope of suppression, maybe we can limit it to the standard library, not all external libraries, which can not be causually changed (I don't know if there is information to judge whether it comes from the standard library). Or just suppress external macros that are not from the current workspace, but I don't know if there is such fine information.

xizheyin commented:
I tried `cargo-bisect-rustc`, and found something interesting.

When I switched the toolchain to `nightly` or `stable`, which are both new versions i.e. `rustc 1.88.0-nightly (092a284ba 2025-04-13)` and `rustc 1.86.0 (05f9846f8 2025-03-31)`), suggestions in the standard library were showing up.

But I switched the toolchain to the specific version, i.e. nightly-2025-04-13, the suggestion was gone.
```
$ rustup default nightly-2025-04-14
info: using existing install for 'nightly-2025-04-14-x86_64-unknown-linux-gnu'
info: default toolchain set to 'nightly-2025-04-14-x86_64-unknown-linux-gnu'

  nightly-2025-04-14-x86_64-unknown-linux-gnu unchanged - rustc 1.88.0-nightly (092a284ba 2025-04-13)
```

The same toolchain, except that one installs via NIGHTLY and the other installs via a specified date, yielding different results.
I need to check the appropriate source code. Do you have any suggestions? @fmease

xizheyin committed:
Suppress suggestions while span is in external library

Signed-off-by: xizheyin <xizheyin@smail.nju.edu.cn>

fmease commented on compiler/rustc_errors/src/emitter.rs:
```diff
@@ -3532,3 +3540,13 @@ pub(crate) fn should_show_source_code(
+fn is_external_library(sm: &SourceMap, span: Span) -> bool {
+    let filename = sm.span_to_filename(span);
+    if let Some(path) = filename.into_local_path() {
+        path.to_string_lossy().contains(".cargo/registry/src/")
+            || path.to_string_lossy().contains(".rustup/toolchains/")
```
Hard-coding these paths is not an option, it's too fragile and too tightly couples us to Cargo and rustup. Re. fragile: The user may reconfigure these paths via the environment variables `CARGO_HOME` and `RUSTUP_HOME`, respectively.

I'll have a think later what alternatives there are (maybe the sysroot path, …).

xizheyin replied:
Yeah, I'm parsing CARGO_HOME and RUSTUP_HOME with environment variables for now, and using Pathbuf to manage paths and avoid platform dependencies. There doesn't seem to be any difference between the different platforms?

Reference:
1. Rustup Book: https://rust-lang.github.io/rustup/environment-variables.html#environment-variables
2. Cargo Book: https://doc.rust-lang.org/cargo/reference/environment-variables.html

We can think about whether there are other situations that have been overlooked.

xizheyin replied:
I also added support for sysroot.
https://github.com/rust-lang/rust/blob/c580c498a1fe144d7c5b2dfc7faab1a229aa288b/src/bootstrap/src/bin/rustc.rs#L58

fmease reviewed (commented):
I'll try to bisect later to see if these issues represent regressions or not.

xizheyin committed:
use env variables to avoid hardcode, and use paltform independent path

Signed-off-by: xizheyin <xizheyin@smail.nju.edu.cn>

xizheyin committed:
Add support for rustc sysroot

Signed-off-by: xizheyin <xizheyin@smail.nju.edu.cn>

jieyouxu commented on compiler/rustc_errors/src/emitter.rs:
```diff
@@ -3532,3 +3540,47 @@ pub(crate) fn should_show_source_code(
+        // use env variable to get path, avoid hardcode
+        // use platform independent path
+
+        let cargo_home = match std::env::var("CARGO_HOME") {
+            Ok(dir) => std::path::PathBuf::from(dir),
+            Err(_) => {
+                if let Ok(home) = std::env::var("HOME") {
+                    std::path::PathBuf::from(home).join(".cargo")
+                } else {
+                    return false;
+                }
+            }
+        };
```
Discussion: this feels really iffy to me, that we're querying so many env vars. E.g. I think on Windows, `HOME` isn't even set. This also magically encodes directory structures that are not guaranteed to remain the same (rustup, cargo, sysroot directory layouts). See for instance #135278 and #135501. Note that this also can't account for remapped paths, I think.

fmease replied:
Yeah, for sure. This shouldn't be the approach we're taking. When I mentioned `RUSTUP_HOME` and `CARGO_HOME` in a previous review comment, I was intending to *dissuade* this kind of approach by demonstrating the fragile nature, the amount of hacks/assumptions and unnecessary coupling to downstream tools.

I haven't re-reviewed yet because I need to think about a non-hacky approach but I first wanted to investigate things.

jieyouxu replied:
Yeah, that's the feeling I had from reading the previous review comments. I mostly wrote my own comment to prevent myself from forgetting 😆

xizheyin replied:
I probably understand. I'm not familiar with this section and may need to investigate further. Are there any other resources I can refer to?

jieyouxu commented:
std sources aren't available in ui tests because we explicitly set `-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX` + `-Ztranslate-remapped-path-to-local-path=no`

https://github.com/rust-lang/rust/blob/6e23095adf9209614a45f7f75fea36dad7b92afb/src/tools/compiletest/src/runtest.rs#L1411-L1420

ehuss commented:
There's some more detail and examples about this in https://github.com/rust-lang/rust/issues/88514. Would be helpful to know if those will all get addressed.

My understanding is that rustc generally already suppresses suggestions for external crates. This may be done around [here](https://github.com/rust-lang/rust/blob/fd50e1012f79eb93f661adc6f276335e4cfb34ea/compiler/rustc_middle/src/lint.rs#L340-L342)? I'm not sure. I'm guessing that this isn't working for suggestions in macros because those diagnostics have their primary spans in the local crate, but the suggestion in an external?

I also thought that `fix_multispan_in_extern_macros` would also have some effect here, but I'm not sure.

fmease commented:
Brief update: We probably want to go through each issue individually and see if instead of suppressing the suggestion altogether we can leverage SyntaxContext functions like `find_oldest_ancestor_in_same_ctxt`, `find_ancestor_in*` to "recover" / generate a non-broken suggestion.
